### PR TITLE
Categorizing Adjectives

### PR DESCRIPTION
**Pull request by guptashashij**
Several methods implemented to categorize adjectives into semantic categories.

### CODE REVIEW

File: categorize_adjectives/libraries/bert.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+from __future__ import absolute_import
```
Comment (sharsh-google) [inline categorize_adjectives/libraries/bert.py:2]: Add a file level comment describing what the file contains.

File: categorize_adjectives/libraries/bert.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+from __future__ import division
```
Comment (sharsh-google) [inline categorize_adjectives/libraries/bert.py:3]: Add function level comments throughout the file, describing what function does and the parameters.

File: categorize_adjectives/libraries/bert.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        self.model = BertModel.from_pretrained(self.bert_model)
+        self.model.to(self.device)
+
+        if n_gpu > 1:
+            self.model = torch.nn.DataParallel(self.model)
+
+        self.model.eval()
+
+
+    def get_sim(self, sentences, layer=-1, metric='cosine'):
```
Comment (sharsh-google) [inline categorize_adjectives/libraries/bert.py]: Rename to get_similarity. Use descriptive names.

File: categorize_adjectives/libraries/bert.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+from scipy.spatial.distance import cosine, euclidean
+
+
+class BertSimilarity(object):
+
+    def __init__(self, bert_model='bert-base-uncased', max_seq_length=50, device='cpu'):
+        self.bert_model = bert_model
+        self.max_seq_length = max_seq_length
+
+        self.device = torch.device("cpu" if device=='cpu' or not torch.cuda.is_available() else "cuda")
```
Comment (sharsh-google) [inline categorize_adjectives/libraries/bert.py]: add space. device == 'cpu'

File: categorize_adjectives/libraries/bert.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+from pytorch_pretrained_bert.tokenization import BertTokenizer
+from pytorch_pretrained_bert.modeling import BertModel
+
+from scipy.spatial.distance import cosine, euclidean
+
+
+class BertSimilarity(object):
+
+    def __init__(self, bert_model='bert-base-uncased', max_seq_length=50, device='cpu'):
```
Comment (sharsh-google) [inline categorize_adjectives/libraries/bert.py:22]: add space. device = 'cpu' 
Here and everywhere

File: categorize_adjectives/libraries/bert.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+            all_encoder_layers, _ = self.model(input_ids, token_type_ids=None, attention_mask=input_mask)
+            all_encoder_layers = all_encoder_layers
+
+            values = torch.mean(all_encoder_layers[layer], 1)
+                
+            out_features.append(values.detach().cpu().numpy())
+
+        flat_list = [item for sublist in out_features for item in sublist]
+        l1 = [e for i, e in enumerate(flat_list) if i %2 == 0]
```
Comment (sharsh-google) [inline categorize_adjectives/libraries/bert.py]: Rename l1, l2 to what they represent

File: categorize_adjectives/using_kmedoids.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
+"""Categorize Adjectives using K-medoids algorithm
```
Comment (sharsh-google) [inline categorize_adjectives/using_kmedoids.py]: The filename should not ever be "using..". It should just be reflective of what it contains. "kmedoids.py" is fine.

File: categorize_adjectives/using_kmedoids.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import numpy as np
+from pyclustering.cluster.kmedoids import kmedoids
+from pyclustering.utils.metric import distance_metric, type_metric
+from libraries.bert import BertSimilarity
+
+# Folder where generated files are stored
+generated_file = os.path.join(os.getcwd(), '..', 'generated_files/')
+
+# Initializing the BERT Similarity function and also storing calculated distances in array.
+x = BertSimilarity()
```
Comment (sharsh-google) [inline categorize_adjectives/using_kmedoids.py]: Rename x to descriptive name

File: categorize_adjectives/keywords.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+# Fetching all adjective synsets from Wordnet
+adjectives = wn.all_synsets('a')
+
+# Semantic Categories for adjectives
+categories = {'Perceptional' : [], 'Spatial' : [], 'Time' : [], 'Motion' : [], 'Substance' : [],
+              'Weather' : [], 'Body' : [], 'Mood' : [], 'Spirit' : [], 'Behavior' : [],
+              'Social' : [], 'Quantity' : [], 'Relational' : [], 'General' : [], 'Others' : []}
+
+# Keywords used to lookout for adjectives
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/keywords.py]: Can you please add a link to the paper here from where we picked up these categories?

File: categorize_adjectives/keywords.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    adj_name = adjective.lemmas()[0].name().lower()
+    similar_synset = wn.synset(adjective.name()).similar_tos()
+    similar_adj = []
+    similar_adj.append(adj_name)
+    # Finding similar adjectives
+    for syn in similar_synset:
+        for lemma in syn.lemmas():
+            name = lemma.name().lower()
+            similar_adj.append(name)
+            if lemma.antonyms():
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/keywords.py]: Why not add synonyms as well?

File: categorize_adjectives/kmedoids.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+from libraries.bert import BertSimilarity
+
+# Folder where generated files are stored
+generated_file = os.path.join(os.getcwd(), '..', 'generated_files/')
+
+# Initializing the BERT Similarity function and also storing calculated distances in array.
+BERT = BertSimilarity()
+dis_array = np.zeros((150, 150)) - 1
+
+def distance(w_1, w_2):
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/kmedoids.py:12]: Please add a comment of what distance is being computed.

File: categorize_adjectives/keywords.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+keywords['Behavior'] = ['behavior', 'character', 'animal specific', 'skill', 'relation', 'sympathy',
+                        'inclination']
+keywords['Social'] = ['stratum', 'social', 'institution', 'poltics', 'religion', 'state', 'region']
+keywords['Quantity'] = ['quantity', 'number', 'cost', 'return']
+keywords['Relational'] = ['relational', 'validity', 'certainty', 'requirement', 'effectiveness',
+                          'difficulty', 'energy requirement', 'functioning', 'security', 'order',
+                          'linking', 'correspondence', 'accuracy', 'completeness', 'cause',
+                          'reference', 'beneficial effect']
+keywords['General'] = ['comparative', 'evaluation', 'standard']
+
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/keywords.py:37]: Please restructure the code below to 3 functions (finding adjectives, dumping the output, main)  and call them from main function.

File: categorize_adjectives/kmedoids.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    if dis_array[w_1, w_2] == -1:
+        dis_array[w_1, w_2] = 1-(BERT.get_similarity([(selected_adj[int(w_1)], selected_adj[int(w_2)])],
+                                 layer=-2)[0])
+        dis_array[w_2, w_1] = dis_array[w_1, w_2]
+    return dis_array[w_1, w_2]
+
+# Fetching the noun to adj dictionary from pickle file
+with open(generated_file + 'noun_to_adj_score.dat', 'wb') as f:
+    noun_to_adj = pickle.load(f)
+
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/kmedoids.py:76]: Please refactor the below code to create multiple functions and add a main function.
Same as above file.

File: categorize_adjectives/libraries/bert.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
+"""Contains function for calculating BERT similarity"""
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/libraries/bert.py:1]: Can you add a link to this source code?

**Review by sowjanyavemuri-google — COMMENTED** [review]
Add a README file for this directory mentioning what each file does.

File: categorize_adjectives/keywords.py (hunk 10 above)
Comment (guptashashij) [reply categorize_adjectives/keywords.py]: Synonyms are covered under similar

File: categorize_adjectives/libraries/bert.py (hunk 14 above)
Comment (guptashashij) [reply categorize_adjectives/libraries/bert.py:1]: This was written by one of my colleague at ETH

File: categorize_adjectives/libraries/bert.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+        if n_gpu > 1:
+            self.model = torch.nn.DataParallel(self.model)
+
+        self.model.eval()
+
+
+    def get_similarity(self, sentences, layer=-1, metric='cosine'):
+        """Returns similarity between sentences
+        """
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/libraries/bert.py]: nit: Move this to above line.

File: categorize_adjectives/keywords.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
+"""Categorize Adjectives by looking for keywords in definition"""
+import os
+from nltk.corpus import wordnet as wn
+
+def find_adjectives(adjectives, categories, keywords):
+"""For each adjective, finding similar adjectives and adding to respective categories"""
```
Comment (sowjanyavemuri-google) [inline categorize_adjectives/keywords.py]: Please update the comments according to this format (go/pystyle#function-docs) i.e add returns, args.
Also, please change the description to have first line as a summary followed by a blank line and then a detailed description.

Same for all methods.

**Review by sowjanyavemuri-google — COMMENTED** [review]
Overall, PR looks good. Thanks!

**Review by sowjanyavemuri-google — APPROVED** [review]

**Review by sharsh-google — APPROVED** [review]

**Pull request merged**